lib/auth/middleware.go

```diff
@@ -103,31 +124,101 @@ func NewTLSServer(cfg TLSServerConfig) (*TLSServer, error) {
 	authMiddleware := &AuthMiddleware{
 		AccessPoint:   cfg.AccessPoint,
 		AcceptedUsage: cfg.AcceptedUsage,
+		Limiter:       limiter,
 	}
-	authMiddleware.Wrap(NewGRPCServer(cfg.APIConfig))
+	authMiddleware.Wrap(NewAPIServer(&cfg.APIConfig))
 	// Wrap sets the next middleware in chain to the authMiddleware
 	limiter.WrapHandle(authMiddleware)
 	// force client auth if given
 	cfg.TLS.ClientAuth = tls.VerifyClientCertIfGiven
 	cfg.TLS.NextProtos = []string{http2.NextProtoTLS}
 
 	server := &TLSServer{
-		TLSServerConfig: cfg,
-		Server: &http.Server{
+		cfg: cfg,
+		httpServer: &http.Server{
 			Handler:           limiter,
 			ReadHeaderTimeout: defaults.DefaultDialTimeout,
 		},
-		Entry: logrus.WithFields(logrus.Fields{
+		log: logrus.WithFields(logrus.Fields{
 			trace.Component: cfg.Component,
 		}),
 	}
-	server.TLS.GetConfigForClient = server.GetConfigForClient
+	server.cfg.TLS.GetConfigForClient = server.GetConfigForClient
+
+	server.grpcServer, err = NewGRPCServer(GRPCServerConfig{
+		TLS:               server.cfg.TLS,
+		APIConfig:         cfg.APIConfig,
+		UnaryInterceptor:  authMiddleware.UnaryInterceptor,
+		StreamInterceptor: authMiddleware.StreamInterceptor,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	server.mux, err = multiplexer.NewTLSListener(multiplexer.TLSListenerConfig{
+		Listener: tls.NewListener(cfg.Listener, server.cfg.TLS),
+		ID:       cfg.ID,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
 	return server, nil
 }
 
-// Serve takes TCP listener, upgrades to TLS using config and starts serving
-func (t *TLSServer) Serve(listener net.Listener) error {
-	return t.Server.Serve(tls.NewListener(listener, t.TLS))
+// Close closes TLS server non-gracefully - terminates in flight connections
+func (t *TLSServer) Close() error {
+	errC := make(chan error, 2)
+	go func() {
+		errC <- t.httpServer.Close()
+	}()
+	go func() {
+		t.grpcServer.server.Stop()
+		errC <- nil
+	}()
+	errors := []error{}
+	for i := 0; i < 2; i++ {
+		errors = append(errors, <-errC)
+	}
+	return trace.NewAggregate(errors...)
```

Comment:
Nit: only one goroutine produces a possible error.  Simpler to just use `chan struct{}` for synchronization and assign directly to an `err` variable (safe because all channel operations are memory barriers).

Reply:
Fair point, but I'd prefer to keep this channel pattern because we use it elsewhere - familiar to the readers